### PR TITLE
add launch function

### PR DESCRIPTION
**Pull request by vchuravy**
bors try

### CODE REVIEW

**Comment by vchuravy** [comment]
@lcw @ali-ramadhan would be good to have your feedback on this design.

``` julia
function div_kernel(::Val{Dev}, f, div_f) where Dev
    @setup Dev

    Nx, Ny, Nz = size(f)
    Δx, Δy, Δz = 1, 1, 1

    # Calculate the divergence of f at every point and store it in div_f.
    @loop for k in (1:Nz; blockIdx().z)
        @loop for j in (1:Ny; (blockIdx().y - 1) * blockDim().y + threadIdx().y)
            @loop for i in (1:Nx; (blockIdx().x - 1) * blockDim().x + threadIdx().x)
                @inbounds div_f[i, j, k] = div(f, Nx, Ny, Nz, Δx, Δy, Δz, i, j, k)
            end
        end
    end

    @synchronize
end

function GPUifyLoops.launch_config(::typeof(div_kernel), maxthreads, dev, f, args...; kwargs...)
   # calculate a good launch config here
   Nx, Ny, Nz = size(f)
   Tx, Ty = 16, 16  # Threads per block
   Bx, By, Bz = Int(Nx/Tx), Int(Ny/Ty), Nz  # Blocks in grid.

  @assert Tx*Ty <= maxthreads
  (threads=(Tx, Ty), blocks=(Bx, By, Bz))
end

calc_div(f::CuArray, div_f::CuArray) = launch(div_kernel, Val(:GPU), f, div_f)
```

The primary idea is to separate the launch configuration from  the actually launch so that we have a chance to query the maximum number of threads
so the we can make choices based on it. I can make more information available to the user, such as `register` usage or `shmem` usage, but maxthreads is a good proxy for that.
`launch` will then query that configuration and use it to schedule the kernel.

**Comment by lcw** [comment]
Seems reasonable to me.  As I understand it, right now `GPUifyLoops.launch_config` is only useful when `dev=Val(:GPU)`.  So should it only be defined in this case?  What does it mean to provide a launch config when `dev=Val(:CPU)`?

**Comment by vchuravy** [comment]
It has no meaning. `launch` should only be called when you run on the GPU and is basically a first step towards fixing #3 since it get's rid of the `@cuda` macro that complicates using Requires.jl

**Comment by ali-ramadhan** [comment]
This looks great and would help clean up the time stepping code we have as there are places where 5-6 kernels are launched after each other and `@hascuda @cuda threads=(Tx, Ty), blocks=(Bx, By, Bz) ...` gets pretty tedious.

Since `::typeof(div_kernel)` is an argument to `launch_config` does that mean that you need to specify a launch config for each kernel? Or can we specify a single _global_ launch config and use it for all kernel launches?

**Comment by vchuravy** [comment]
I just noticed that I forgot to push the prototype. You could provide a global launch configuration, but  that means overwriting the current default, which allows:

```
launch(kernel, args...; threads=(64, 64, 1))
```

So I would stick with local configurations. Sadly we can't have a dispatch hierarchy for functions. Although we might be able to play some tricks with modules.

**Comment by vchuravy** [comment]
So one thing I am unhappy about is that we lose the ability to pass keyword arguments to the kernel. So maybe a macro `@launch` would be better

**Comment by glwagner** [comment]
Is it always possible to calculate a good launch configuration / is there a way to make the launch configuration settable by the user? Could it make sense to store launch configurations in the `CUDA` `Device` (default or 'good' configurations being calculated on construction) instead? Forgive me for bumbling / misunderstanding...

**Comment by vchuravy** [comment]
You can always specify a local launch configuration,.

```
function div_kernel(::Val{Dev}, f, div_f) where Dev
    @setup Dev

    Nx, Ny, Nz = size(f)
    Δx, Δy, Δz = 1, 1, 1

    # Calculate the divergence of f at every point and store it in div_f.
    @loop for k in (1:Nz; blockIdx().z)
        @loop for j in (1:Ny; (blockIdx().y - 1) * blockDim().y + threadIdx().y)
            @loop for i in (1:Nx; (blockIdx().x - 1) * blockDim().x + threadIdx().x)
                @inbounds div_f[i, j, k] = div(f, Nx, Ny, Nz, Δx, Δy, Δz, i, j, k)
            end
        end
    end

    @synchronize
end

function calc_div(f::CuArray, div_f::CuArray)
   # calculate a good launch config here
   Nx, Ny, Nz = size(f)
   Tx, Ty = 16, 16  # Threads per block
   Bx, By, Bz = Int(Nx/Tx), Int(Ny/Ty), Nz  # Blocks in grid.

  launch(div_kernel, Val(:GPU), f, div_f; threads=(Tx, Ty), blocks=(Bx, By, Bz))
end
```

**Comment by vchuravy** [comment]
I have changed this to a macro that also applies `contextualize` to the kernel.

bors try

**Comment by vchuravy** [comment]
bors try

**Comment by vchuravy** [comment]
@lcw, @glwagner and @charleskawczynski 

This is pretty close to what I envisioned initially. This gets rid of the need for the setup call, and the second layer of dispatch.
There is still an odd bug that I need to chase down, but I would like to feedback before the interface becomes to settle.

**Comment by lcw** [comment]
It is looking good to me!

What happens if I call `@launch CPU() threads=length(A) kernel!(A, B, f)`?  Are the kwargs ignored?  Or do we want to use `launch_config` and then the generic way of launching a kernel would be `@launch DEV() kernel!(A, B, f)` where `DEV()` could be `CPU()`, `CUDA()`, `ROCm()`, ...?

**Comment by vchuravy** [comment]
Yeah the kwargs get discarded (see https://github.com/vchuravy/GPUifyLoops.jl/pull/20/files#diff-274be86b324412b14cfc4303c5c8ab1cR59). It might become interesting when we add a multi-threaded CPU backend.

File: src/GPUifyLoops.jl [hunk 1]
```diff
@@ -20,54 +20,98 @@ export @setup, @loop, @synchronize
...
+            config = launch_config(f, maxthreads, args...; call_kwargs)
+
+            kernel(kernel_args...; config...)
+        end
+        return nothing
+    end
 end
 
+isdevice() = false
+Cassette.overdub(ctx::Ctx, ::typeof(isdevice)) = true
```
Comment (glwagner) [inline src/GPUifyLoops.jl:117]: Forgive me for my ignorance, but what happens here? This function is changed to output true if the code is compiled for GPU, somehow?

File: src/GPUifyLoops.jl [hunk 2]
```diff
@@ -20,54 +20,98 @@ export @setup, @loop, @synchronize
...
 
 """
-    @setup Dev
+   launch(::Device, f, args..., kwargs...)
 
-Setups some hidden state within the function that allows the other macros to
-properly work.
+Launch a kernel on the GPU. `kwargs` are passed to `@cuda`
+`kwargs` can be any of the compilation and runtime arguments
+normally passed to `@cuda`.
```
Comment (glwagner) [inline src/GPUifyLoops.jl:57]: Maybe good to mention here that `launch` does nothing on the CPU?

File: src/GPUifyLoops.jl (hunk 1 above)
Comment (vchuravy) [reply src/GPUifyLoops.jl:117]: Yes basically. We use Cassette for contextual execution and within a contexts it is going to replace the original function with this one

**Comment by vchuravy** [comment]
Bors try

**Comment by vchuravy** [comment]
bors try

**Comment by vchuravy** [comment]
bors try

**Comment by vchuravy** [comment]
bors try

**Comment by vchuravy** [comment]
bors try

**Comment by vchuravy** [comment]
and again bors try, I should do my dev work on a GPU enabled machine...

**Pull request merged**